# What is a Decision Tree?

<b>Decision tree </b> is a type of <b>supervised learning algorithm</b> that is mostly used in classification problems. It works for both categorical and continuous input and output variables. 

Example:-

Let’s say we have a sample of <b>30 students </b>with three variables Gender (Boy/Girl), Class(IX/X) and Height (5 to 6 ft). 15 out of these 30 play cricket in leisure time. Now, I want to create a model to predict who will play cricket during leisure period? In this problem, we need to segregate students who play cricket in their leisure time based on highly significant input variable among all three.

This is where decision tree helps, it will segregate the students based on all values of three variable and identify the variable, which creates the best homogeneous sets of students (which are heterogeneous to each other). In the snapshot below, you can see that variable Gender is able to identify best homogeneous sets compared to the other two variables.



# Practical Implementation

The file **daily_weather.csv** is a comma-separated file that contains weather data.    The weather station is equipped with sensors that capture weather-related measurements such as air temperature, air pressure, and relative humidity.  Data was collected for a period of three years, from September 2011 to September 2014, to ensure that sufficient data for different seasons and weather conditions is captured.

**Problem: Use morning sensor signals as features to predict whether the humidity will be high at 3pm.**


First we import the necessary libraries of the python for demostration of the Decision Tree Classifier.

In [ ]:
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

Read the data of the weather from the csv file using read_csv function of pandas dataframe.

In [ ]:
data = pd.read_csv('daily_weather.csv')

In [ ]:
data.head()

air_pressure_9am  air_temp_9am  avg_wind_direction_9am  avg_wind_speed_9am  \
0        918.060000     74.822000              271.100000            2.080354   
1        917.347688     71.403843              101.935179            2.443009   
2        923.040000     60.638000               51.000000           17.067852   
3        920.502751     70.138895              198.832133            4.337363   
4        921.160000     44.294000              277.800000            1.856660   

   max_wind_direction_9am  max_wind_speed_9am  rain_accumulation_9am  \
0              295.400000            2.863283                    0.0   
1              140.471549            3.533324                    0.0   
2               63.700000           22.100967                    0.0   
3              211.203341            5.190045                    0.0   
4              136.500000            2.863283                    8.9   

   rain_duration_9am  relative_humidity_9am  high_humidity_3pm  
0                0.0              42.420000                  1  
1                0.0              24.328697                  0  
2               20.0               8.900000                  0  
3                0.0              12.189102                  0  
4            14730.0              92.410000                  1

In [ ]:
data.columns

Index(['air_pressure_9am', 'air_temp_9am', 'avg_wind_direction_9am',
       'avg_wind_speed_9am', 'max_wind_direction_9am', 'max_wind_speed_9am',
       'rain_accumulation_9am', 'rain_duration_9am', 'relative_humidity_9am',
       'high_humidity_3pm'],
      dtype='object')

In [ ]:
data.shape

(1095, 10)

In [ ]:
data.describe()

air_pressure_9am  air_temp_9am  avg_wind_direction_9am  \
count       1092.000000   1090.000000             1091.000000   
mean         918.882551     64.933001              142.235511   
std            3.184161     11.175514               69.137859   
min          907.990000     36.752000               15.500000   
25%          916.550000     57.281000               65.972506   
50%          918.921045     65.715479              166.000000   
75%          921.160073     73.450974              191.000000   
max          929.320000     98.906000              343.400000   

       avg_wind_speed_9am  max_wind_direction_9am  max_wind_speed_9am  \
count         1092.000000             1092.000000         1091.000000   
mean             5.508284              148.953518            7.019514   
std              4.552813               67.238013            5.598209   
min              0.693451               28.900000            1.185578   
25%              2.248768               76.553003            3.067477   
50%              3.871333              177.300000            4.943637   
75%              7.337163              201.233153            8.947760   
max             23.554978              312.200000           29.840780   

       rain_accumulation_9am  rain_duration_9am  relative_humidity_9am  \
count            1089.000000        1092.000000            1095.000000   
mean                0.203079         294.108052              34.241402   
std                 1.593952        1598.078779              25.472067   
min                 0.000000           0.000000               6.090000   
25%                 0.000000           0.000000              15.092243   
50%                 0.000000           0.000000              23.179259   
75%                 0.000000           0.000000              45.400000   
max                24.020000       17704.000000              92.620000   

       high_humidity_3pm  
count        1095.000000  
mean            0.499543  
std             0.500228  
min             0.000000  
25%             0.000000  
50%             0.000000  
75%             1.000000  
max             1.000000

In [ ]:
data['high_humidity_3pm'].value_counts()

0    548
1    547
Name: high_humidity_3pm, dtype: int64

Removing the rows which contains the null values

In [ ]:
data.isnull().sum()

air_pressure_9am          3
air_temp_9am              5
avg_wind_direction_9am    4
avg_wind_speed_9am        3
max_wind_direction_9am    3
max_wind_speed_9am        4
rain_accumulation_9am     6
rain_duration_9am         3
relative_humidity_9am     0
high_humidity_3pm         0
dtype: int64

In [ ]:
data.dropna(inplace=True)

In [ ]:
data.shape

(1064, 10)

In [ ]:
independent_variables = ['air_pressure_9am', 'air_temp_9am', 'avg_wind_direction_9am',
       'avg_wind_speed_9am', 'max_wind_direction_9am', 'max_wind_speed_9am',
       'rain_accumulation_9am', 'rain_duration_9am', 'relative_humidity_9am']

In [ ]:
dependent_variable = 'high_humidity_3pm'

In [ ]:
X = data.drop(columns='high_humidity_3pm',axis=1)
Y = data['high_humidity_3pm']

By using train_test_split we have split the data into traing dataset and testing datasets.

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.33, random_state=112)

**Fit on Train Set**

In [ ]:
dec_classifier = DecisionTreeClassifier(criterion='gini', max_leaf_nodes=20)
dec_classifier.fit(X_train, Y_train)

DecisionTreeClassifier(max_leaf_nodes=20)

In [ ]:
y_pred_train = dec_classifier.predict(X_train)

In [ ]:
accuracy_score(Y_train,y_pred_train)

0.9353932584269663

**Predict on Test Set**

In [ ]:
y_predicted = humidity_classifier.predict(X_test)

In [ ]:
y_predicted[:10]

array([1, 1, 0, 1, 0, 0, 0, 0, 1, 0], dtype=int64)

In [ ]:
y_test[:10]

835    1
910    1
646    0
350    1
2      0
797    0
423    1
850    0
302    1
284    0
Name: high_humidity_3pm, dtype: int64

### Measure Accuracy of the Classifier

In [ ]:
accuracy_score(y_test, y_predicted) * 100

88.63636363636364

The conclusion is that the model has good accuracy in predicting whether the humidity will be high at 3pm using morning sensor signals as features. The accuracy on the training set is 93.54%, and the accuracy on the test set is 88.64%, which is a good indication that the model generalizes well to unseen data.